opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java

```diff
@@ -0,0 +1,78 @@
+import de.adorsys.opba.protocol.xs2a.service.xs2a.dto.oauth2.Xs2aOauth2Headers;
+import de.adorsys.opba.protocol.xs2a.service.xs2a.dto.oauth2.Xs2aOauth2WithCodeParameters;
+import de.adorsys.opba.protocol.xs2a.util.logresolver.Xs2aLogResolver;
+import de.adorsys.xs2a.adapter.api.EmbeddedPreAuthorisationService;
+import de.adorsys.xs2a.adapter.api.model.AspspScaApproach;
+import de.adorsys.xs2a.adapter.api.model.EmbeddedPreAuthorisationRequest;
+import de.adorsys.xs2a.adapter.api.model.TokenResponse;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.flowable.engine.delegate.DelegateExecution;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service("xs2aEmbeddedPreAuthorization")
+@RequiredArgsConstructor
+public class Xs2aEmbeddedPreAuthorization extends ValidatedExecution<Xs2aContext> {
+
+    private final Extractor extractor;
+    private final EmbeddedPreAuthorisationService embeddedPreAuthorisationService;
+    private final Xs2aLogResolver logResolver = new Xs2aLogResolver(getClass());
+
+    @Override
+    @SneakyThrows
+    protected void doRealExecution(DelegateExecution execution, Xs2aContext context) {
+        logResolver.log("doRealExecution: execution ({}) with context ({})", execution, context);
+
+        ValidatedQueryHeaders<Xs2aOauth2WithCodeParameters, Xs2aOauth2Headers> validated = extractor.forExecution(context);
+
+        logResolver.log("getToken with parameters: {}", validated.getQuery(), validated.getHeaders());
+        EmbeddedPreAuthorisationRequest embeddedPreAuthorisationRequest = new EmbeddedPreAuthorisationRequest();
+        embeddedPreAuthorisationRequest.setPassword(context.getPsuPassword());
+        embeddedPreAuthorisationRequest.setUsername(context.getPsuId());
+        TokenResponse response = this.embeddedPreAuthorisationService.getToken(embeddedPreAuthorisationRequest, validated.getHeaders().toHeaders());
+        if (response.getTokenType() == null) {
+            response.setTokenType("Bearer");
+        }
+        logResolver.log("getToken response: {}", response);
+            ContextUtil.getAndUpdateContext(
+                    execution,
+                    (Xs2aContext ctx) -> {
+                        ctx.setOauth2Token(response);
+                        ctx.setEmbeddedPreAuthNeeded(false);
+                        ctx.setEmbeddedPreAuthDone(true);
+
+                        log.info("aspsp sca approach: {}", ctx.getAspspScaApproach());
+                        ctx.setAspspScaApproach(AspspScaApproach.EMBEDDED.name());
+                    }
+            );
+
+
+    }
+
+    @Override
+    protected void doValidate(DelegateExecution execution, Xs2aContext context) {
+        logResolver.log("doValidate: execution ({}) with context ({})", execution, context);
+
+     //   validator.validate(execution, context, this.getClass(), extractor.forValidation(context));
```

Comment:
Also, according to https://github.com/adorsys/open-banking-gateway/pull/1284/files#r682317167
suggest to request PSU password/login here